dojo_plugin/api/v1/dojo.py

```diff
@@ -0,0 +1,244 @@
+                assert len(dojo_specs) == 1, (
+                    f"Dojo repository must have exactly one top-level dojo spec yml named {{YOUR_DOJO_ID}}.yml. Yours has: {dojo_specs}"
+                )
+                dojo_id = dojo_specs[0].stem
+                assert id_regex(dojo_id), (
+                    f"Your dojo ID (the extensionless name of your .yml file) must be a valid URL component."
+                )
+
+                # make sure there aren't any symlink shenanigans
+                assert dojo_specs[0].is_file() and not dojo_specs[0].is_symlink(), (
+                    f"{dojo_specs[0].name} is not a regular file!"
+                )
+
+                # make sure we're not overwriting unintentionally
+                dojo_permanent_dir = pathlib.Path(DOJOS_DIR)/str(user.id)/dojo_id
+                if data.get("dojo_replace", False) not in ("true", True, 1):
+                    assert not dojo_permanent_dir.exists(), (
+                        f"You already have a cloned dojo repository containing a dojo with ID {dojo_id}."
+                    )
+
+                # make sure we're not overwriting someone else's dojo
+                existing_dojo = Dojos.query.filter_by(id=dojo_id).first()
+                if existing_dojo is not None:
+                    assert existing_dojo.owner_id == user.id, (
+                        f"A dojo with the ID {dojo_id} was already created by a different user. Please choose a different ID."
+                    )
+                    join_code = existing_dojo.join_code
+                else:
+                    join_code = random_dojo_join_code()
+
+                # do a test load
+                log_handler = HTMLHandler()
+                logger = logging.getLogger(f"dojo-load-{user.id}-{dojo_id}")
+                logger.setLevel('DEBUG')
+                logger.addHandler(log_handler)
+                load_dojo(
+                    dojo_id, dojo_specs[0].read_text(),
+                    user=user, commit=False, dojo_dir=clone_dir, log=logger
+                )
+                assert ("WARNING" not in log_handler.html) and ("ERROR" not in log_handler.html), (
+                    "A test load of your dojo resulted in the following log messages. Please fix all warnings and try again.<br>" +
+                    log_handler.html
+                )
+
+                # move the pulled dojo in
+                if dojo_permanent_dir.exists():
+                    shutil.rmtree(dojo_permanent_dir)
+                dojo_permanent_dir.parent.mkdir(parents=True, exist_ok=True)
+                shutil.move(clone_dir, dojo_permanent_dir)
+
+                # load it for real!
+                log_handler.reset()
+                load_dojo(
+                    dojo_id, (dojo_permanent_dir/(dojo_id+".yml")).read_text(),
+                    user=user, commit=True, dojo_dir=dojo_permanent_dir, log=logger, initial_join_code=join_code
```

Comment:
We should make `join_code`s specified in the dojo yml.
We should make public vs private specified in the dojo yml (default to private).

Reply:
I like this idea (as long as we do the necessary check here for public dojos).